src/main/java/org/cryptomator/linux/quickaccess/DolphinPlaces.java

```diff
@@ -0,0 +1,161 @@
+		String id = UUID.randomUUID().toString();
+		try {
+			MODIFY_LOCK.lock();
+			if (Files.size(PLACES_FILE) > MAX_FILE_SIZE) {
+				throw new IOException("File %s exceeds size of %d bytes".formatted(PLACES_FILE, MAX_FILE_SIZE));
+			}
+			var placesContent = Files.readString(PLACES_FILE);
+			//validate
+			xmlValidator.validate(new StreamSource(new StringReader(placesContent)));
+			// modify
+			int insertIndex = placesContent.lastIndexOf("</xbel>"); //cannot be -1 due to validation
+			try (var writer = Files.newBufferedWriter(TMP_FILE, StandardCharsets.UTF_8, StandardOpenOption.WRITE, StandardOpenOption.CREATE, StandardOpenOption.TRUNCATE_EXISTING)) {
+				writer.write(placesContent, 0, insertIndex);
+				writer.newLine();
+				writer.write(ENTRY_TEMPLATE.formatted(target.toUri(), displayName, id).indent(1));
+				writer.newLine();
+				writer.write(placesContent, insertIndex, placesContent.length() - insertIndex);
+			}
+			// save
+			Files.move(TMP_FILE, PLACES_FILE, StandardCopyOption.REPLACE_EXISTING, StandardCopyOption.ATOMIC_MOVE);
+			return new DolphinPlacesEntry(id);
+		} catch (SAXException | IOException e) {
+			throw new QuickAccessServiceException("Adding entry to KDE places file failed.", e);
+		} finally {
+			MODIFY_LOCK.unlock();
+		}
+	}
+
+	private static class DolphinPlacesEntry implements QuickAccessEntry {
+
+		private final String id;
+		private volatile boolean isRemoved = false;
+
+		DolphinPlacesEntry(String id) {
+			this.id = id;
+		}
+
+		@Override
+		public void remove() throws QuickAccessServiceException {
+			try {
+				MODIFY_LOCK.lock();
+				if (isRemoved) {
+					return;
+				}
+				if (Files.size(PLACES_FILE) > MAX_FILE_SIZE) {
+					throw new IOException("File %s exceeds size of %d bytes".formatted(PLACES_FILE, MAX_FILE_SIZE));
+				}
+				var placesContent = Files.readString(PLACES_FILE);
+				int idIndex = placesContent.lastIndexOf(id);
+				if (idIndex == -1) {
+					isRemoved = true;
+					return; //we assume someone has removed our entry
+				}
+				//validate
+				xmlValidator.validate(new StreamSource(new StringReader(placesContent)));
+				//modify
+				var placesContentPart1 = placesContent.substring(0, idIndex);
+				int openingTagIndex = placesContentPart1.lastIndexOf("<bookmark href=");
+				var contentToWrite1 = placesContentPart1.substring(0, openingTagIndex).stripTrailing();
+
+				int closingTagIndex = placesContent.indexOf("</bookmark>", idIndex);
+				var part2Tmp = placesContent.substring(closingTagIndex + "</bookmark>".length()).split("\\v*", 2); //removing leading vertical whitespaces
+				var contentToWrite2 = part2Tmp.length == 1 ? part2Tmp[0] : part2Tmp[1];
+
+				try (var writer = Files.newBufferedWriter(TMP_FILE, StandardCharsets.UTF_8, StandardOpenOption.WRITE, StandardOpenOption.CREATE, StandardOpenOption.TRUNCATE_EXISTING)) {
+					writer.write(contentToWrite1);
+					writer.newLine();
+					writer.write(contentToWrite2);
+				}
+				// save
```

Comment:
Same here: I'd suggest validating our modified file a second time before moving it.

Reply:
WIth commits  5e9eacc5cbd51c4dc3b1ea8a9d5734ad591c49fb  and a5dca5b6390f5bbeb284d76d41ba435fcf7ba493 I improved the match logic to also work for reformatted documents. This ensures, that the result is still a valid xbel document.